### PR TITLE
implemented more operators and some more tests

### PR DESCRIPTION
**Pull request by thomasfanell**
* implemented more comparison operators and operator []
* implemented conjugate property and toArray()
* added more tests
* fixed a minor error in a comment in FloatTransform3D

### CODE REVIEW

**Comment by davidhesselbom** [comment]
`QuaternionTest` is still failing on Travis:
```
Quaternion ..........f... failed

Failed test: ./test/math/QuaternionTest.ooc
```
As before, it looks like `LogarithmExponential` is where things go wrong. The test framework currently doesn't let us see exactly which `expect` causes the test to fail, so can you please try to divide that section up into smaller pieces?

File: source/math/FloatTransform3D.ooc [hunk 1]
```diff
@@ -20,7 +20,7 @@ import FloatPoint3D
 import text/StringTokenizer
 import structs/ArrayList
 
-// The 2D transform is a 3x3 homogeneous coordinate matrix.
+// The 3D transform is a 4x3 homogeneous coordinate matrix.
```
Comment (davidhesselbom) [inline source/math/FloatTransform3D.ooc]: Oh, dear. Thanks!

**Comment by thomasfanell** [comment]
I'll split the test up and see if I can see what the problem is.

File: source/math/Quaternion.ooc [hunk 2]
```diff
@@ -210,8 +224,51 @@ Quaternion: cover {
...
+			case 2 => result = this y
+			case 3 => result = this z
+			case => raise("Invalid index: #{index}, valid indices are 0-3.")
+		}
+		result
+	}
+	operator as -> String { this toString() }
+	toArray: func -> Float[] {
+		result := [this w, this x, this y, this z]
+		result
```
Comment (davidhesselbom) [inline source/math/Quaternion.ooc]: `result` is useful in multiple return functions, but here, it just adds an extra line.
```ooc
	toArray: func -> Float[] {
		[this w, this x, this y, this z]
```
should suffice.

Comment (thomasfanell) [inline source/math/Quaternion.ooc]: Tried that actually, did not compile. Hence the temp variable

File: test/math/QuaternionTest.ooc [hunk 3]
```diff
@@ -28,12 +47,35 @@ QuaternionTest: class extends Fixture {
 		this add("norm", func() {
 			expect(this quaternion0 norm, is equal to(65.5991592f))
 		})
+		this add("conjugate", func() {
+			conjugate := quaternion5 conjugate
+			result := conjugate w == quaternion5 w &&
+				conjugate[1] == -quaternion5[1] &&
+				conjugate[2] == -quaternion5[2] &&
+				conjugate[3] == -quaternion5[3]
+			expect(result)
```
Comment (davidhesselbom) [inline test/math/QuaternionTest.ooc]: Wouldn't this be better as 4 `expect`s?

File: source/math/Quaternion.ooc [hunk 4]
```diff
@@ -210,8 +224,51 @@ Quaternion: cover {
 		This new(realResult, imaginaryResult)
 	}
 	operator * (value: FloatPoint3D) -> FloatPoint3D {
-		(this * Quaternion new (0.0f, value) * this inverse) imaginary
+		(this * This new (0.0f, value) * this inverse) imaginary
+	}
+	operator [] (index: Int) -> Float {
+		result: Float
+		match(index) {
```
Comment (davidhesselbom) [inline source/math/Quaternion.ooc]: `match` is a keyword, so there should be a space between it and `(`.

File: test/math/QuaternionTest.ooc [hunk 5]
```diff
@@ -8,11 +8,30 @@ QuaternionTest: class extends Fixture {
...
+			expect(this quaternion0 <= this quaternion0)
+			expect(this quaternion0 >= this quaternion0)
+			expect(this quaternion6 < this quaternion5)
+			expect(this quaternion5 > this quaternion6)
+		})
+		this add("indexing", func() {
+			expect(this quaternion0[0] == 33.0f)
+			expect(this quaternion0[1] == 10.0f)
+			expect(this quaternion0[2] == -12.0f)
+			expect(this quaternion0[3] == 54.5f)
```
Comment (davidhesselbom) [inline test/math/QuaternionTest.ooc]: Also add a try - catch for indexing out of bounds.

File: source/math/Quaternion.ooc [hunk 6]
```diff
@@ -210,8 +224,51 @@ Quaternion: cover {
...
+			case 0 => result = this w
+			case 1 => result = this x
+			case 2 => result = this y
+			case 3 => result = this z
+			case => raise("Invalid index: #{index}, valid indices are 0-3.")
```
Comment (davidhesselbom) [inline source/math/Quaternion.ooc]: It might be helpful to mention, in the error message, in which class the exception occurred.

File: source/math/Quaternion.ooc (hunk 2 above)
Comment (davidhesselbom) [inline source/math/Quaternion.ooc]: Interesting. I'll give it a try.

File: test/math/QuaternionTest.ooc (hunk 5 above)
Comment (thomasfanell) [inline test/math/QuaternionTest.ooc]: Do you mean I should index out of bounds and catch it?

File: source/math/FloatTransform3D.ooc (hunk 1 above)
Comment (davidhesselbom) [inline source/math/FloatTransform3D.ooc]: Next time, though, this type of thing should be in its own commit, because it's unrelated to the rest.

File: test/math/QuaternionTest.ooc (hunk 5 above)
Comment (davidhesselbom) [inline test/math/QuaternionTest.ooc]: Yes, make sure the exception is thrown as desired.

Comment (thomasfanell) [inline test/math/QuaternionTest.ooc]: I added this test in the amended version

File: source/math/Quaternion.ooc (hunk 2 above)
Comment (davidhesselbom) [inline source/math/Quaternion.ooc]: I'm pretty sure this was fixed in https://github.com/fasterthanlime/rock/commit/1b88d707e8959bcba9da34a8cba5267f4b0e5f7b, but our version of rock doesn't have that fix.
